### ISSUE TITLE
Extend OAuth/OpenID Connect solution to support logout

### ISSUE THREAD

**Issue by Kavindu-Dodan** [keptn/keptn#3880]
**Background**

#3448 focused on supporting user login using OAuth/OpenID Connect. #3780 provided Keptn bridge interface and API contract to implement this.

Logout implemented through #3780 merely focus on ending bridge session. From OpenID Connect perspective, it is desired to see a logout for Authorization Server (aka OpenID Provider - OP)

**Considerations**

Similar to #3448 , the solution must be vendor neutral. From Keptn bridge point of view, we must only fulfil an API contract, facilitated by a supportive service. 

**Solution**

OIDC RP initiated logout [1] is the most straightforward implementation bridge can easily support. This can be built on top of the OAuth service proposed trough #3448. 

Logout interaction can be simplified as below

1. If OAuth service support logout, it can publish through discovery document (i.e - logout_endpoint ) 
2. If logout at OAuth service is available, when bridge perform the logout, bridge can invoke the logout_endpoint. 
3. OAuth service can then generate the logout request for the specific configurations (Note - id_token_hint can be an optional input)
4. Bridge simply redirect the user agent to the specific logout request 

Specification recommends to utilise id_token_hint [1] and for the first iteration we can support this parameter to be stored against user session in bridge. This makes interactions with OAuth service stateless (except for state usage in login flow).

[1] https://openid.net/specs/openid-connect-rpinitiated-1_0.html

**Comment by Kavindu-Dodan** [comment]
This is my proposal for logout API contract.

**TLDR** : No breaking changes & updated swagger documentation - [KeptnOAuthSwaggerV2.txt](https://github.com/keptn/keptn/files/6403229/KeptnOAuthSwaggerV2.txt)

OAuth service have the option to expose a logout endpoint, which closely follow OIDC RP-Initiated Logout [1]. Further, similar to OIDC Discovery [2], support for logout is communicated through OAuth service discovery.

**Discovery**

If OAuth service support RP logout, the discovery response must communicate this.

Keptn bridge looks for the _rp_logout_ key when initialising OAuth middleware. If this is present, keptn bridge will include RP logout flow when user logout.

Given below is a discovery response with _rp_logout_ key, 
```
Get http://oauth-service:8080/discovery

{
 "authorization": "http://oauth-service:8080/authorization",
 "token_decision": "http://oauth-service:8080/token_decision",
 "rp_logout": "http://oauth-service:8080/logout"
}
```

**Logout hint**

When OAuth service support RP logout, it has the option to pass a logout hint in token decision response. If present, keptn bridge will include this hint in the RP logout request. 

Given below is token decision response including a logout hint,
```
Post http://oauth-service:8080/token_decision
Content-Type: application/json
{
  "code": "qwert",
  "state": "123"
}

HTTP 200 OK
Content-Type: application/json
{
  "user": "USER_NAME",
   logout_hint": "LOGOUT_HINT"
}
```

The hint can be anything that require to generate the OIDC RP logout request. For example, this can be the ID Token, which can be linked to _id_token_hint_ parameter consumed by the IDP's end session endpoint. 

**Logout request**

When user logout from Keptn bridge & if OAuth service support logout, bridge will invoke the _rp_logout_ endpoint. Further, this request will include the _logout_hint_ if it was present in the token decision response. 

Response to the this request must contain the _logout_path_. This must be a valid URL & keptn bridge will redirect user agent to this URL.

Given below is a sample request, including logout hint and the response,

```
Post http://oauth-service:8080/rp_logout
Content-Type: application/json
{
  "logout_hint": "s784Opsf320..."
}

HTTP 200 OK
Content-Type: application/json

{
  "logout_path": "http://idp.com/logout?id_token_hint=s784Opsf320..."
}
```


[1] https://openid.net/specs/openid-connect-rpinitiated-1_0.html
[2] https://openid.net/specs/openid-connect-discovery-1_0.html
